### ISSUE TITLE
Bug: "additional options to rsync" not used when removing snapshots

### ISSUE THREAD

**Issue by xelvinc** [bit-team/backintime#1861]
Will be fixed via #2141 but depends on #1984.

---

Short: Additional rsync options (Expert Options -> Paste additional options to rsync) not used when removing snapshots.

When I want to remove a snapshot, I get the error:
```
WARNING: Command "rsync -a --delete -s /tmp/tmps8asu05e/ /mnt/databackup/backintime/backintime/gentoo/root/1/20240502-194442-488" returns 22 | [generator] exceeded --max-alloc=1073741824 setting (file=ifuncs.h, line=30)  
rsync error: error allocating core memory buffers (code 22) at util2.c(80) [generator=3.3.0]  
ERROR: Last rsync command failed with return code "22". See previous WARNING message in the logs for details.
```
According to the rsync man page (I'm using version 3.3.0), the default value for --max-alloc (default is 1g) can be overridden with the --max-alloc=SIZE option.

But when I set this option under Expert Options -> Paste additional options to rsync -> --max-alloc=5g, this option is only passed to the command line when taking a new snapshot. Not when removing a snapshot.
Looking in snapshots.py, the rsync_prefix (which includes `if config.rsyncOptionsEnabled(): cmd.extend(shlex.split(config.rsyncOptions()))`) is only set in the takeSnapshot routine, not in the remove routine.
Shouldn't the rsyncOptions also be used in the remove routine? This was already requested in https://github.com/bit-team/backintime/issues/1507, but for another reason.

How can I work around this? Is it safe to manually run the `rsync -a --delete -s /tmp/tmps8asu05e/ /mnt/databackup/backintime/backintime/gentoo/root/1/20240502-194442-488` command with the `--max-alloc=5g` option? Should I then create an empty directory /tmp/tmps8asu05e/ first?
Or just remove the `/mnt/databackup/backintime/backintime/gentoo/root/1/20240502-194442-488` directory?

I've also set the RSYNC_MAX_ALLOC environment variable to 5g (as described in the rsync man page), but this doesn't make a difference.

```
backintime --diagnostics  
{  
   "backintime": {  
       "name": "Back In Time",  
       "version": "1.3.3",  
       "latest-config-version": 6,  
       "local-config-file": "/home/UsernameReplaced/.config/backintime/config",  
       "local-config-file-found": false,  
       "global-config-file": "/etc/backintime/config",  
       "global-config-file-found": false,  
       "started-from": "/usr/share/backintime/common",  
       "running-as-root": false,  
       "user-callback": "/home/UsernameReplaced/.config/backintime/user-callback",  
       "keyring-supported": false  
   },  
   "host-setup": {  
       "platform": "Linux-6.10.6-gentoo-dist-x86_64-AMD_Ryzen_9_7900X_12-Core_Processor-with-glibc2.39",  
       "system": "Linux #1 SMP PREEMPT_DYNAMIC Tue Aug 20 18:54:03 CEST 2024",  
       "os-release": {  
           "NAME": "Gentoo",  
           "ID": "gentoo",  
           "PRETTY_NAME": "Gentoo Linux",  
           "ANSI_COLOR": "1;32",  
           "HOME_URL": "https://www.gentoo.org/",  
           "SUPPORT_URL": "https://www.gentoo.org/support/",  
           "BUG_REPORT_URL": "https://bugs.gentoo.org/",  
           "VERSION_ID": "2.15"  
       },  
       "display-system": "wayland",  
       "locale": "en_US, UTF-8",  
       "PATH": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/bin:/usr/lib/llvm/18/bin",  
       "RSYNC_OLD_ARGS": "(not set)",  
       "RSYNC_PROTECT_ARGS": "(not set)"  
   },  
   "python-setup": {  
       "python": "3.12.3 main Jun  8 2024 10:56:55 CPython GCC 13.2.1 20240210",  
       "python-executable": "/usr/lib/python-exec/python3.12/python3",  
       "python-executable-symlink": true,  
       "python-executable-resolved": "/usr/bin/python3.12",  
       "sys.path": [  
           "/usr/share/backintime/qt/plugins",  
           "/usr/share/backintime/common/plugins",  
           "/usr/share/backintime/plugins",  
           "/usr/share/backintime/common",  
           "/usr/lib/python312.zip",  
           "/usr/lib/python3.12",  
           "/usr/lib/python3.12/lib-dynload",  
           "/usr/lib/python3.12/site-packages"  
       ],  
       "qt": "PyQt 5.15.11 / Qt 5.15.14"  
   },  
   "external-programs": {  
       "rsync": {  
           "program": "rsync",  
           "version": "3.3.0",  
           "protocol": "31.0",  
           "copyright": "(C) 1996-2024 by Andrew Tridgell, Wayne Davison, and others.",  
           "url": "https://rsync.samba.org/",  
           "capabilities": {  
               "file_bits": 64,  
               "inum_bits": 64,  
               "timestamp_bits": 64,  
               "long_int_bits": 64,  
               "socketpairs": true,  
               "symlinks": true,  
               "symtimes": true,  
               "hardlinks": true,  
               "hardlink_specials": true,  
               "hardlink_symlinks": true,  
               "IPv6": true,  
               "atimes": true,  
               "batchfiles": true,  
               "inplace": true,  
               "append": true,  
               "ACLs": true,  
               "xattrs": true,  
               "secluded_args": "optional",  
               "iconv": true,  
               "prealloc": true,  
               "stop_at": true,  
               "crtimes": false  
           },  
           "optimizations": {  
               "SIMD_roll": true,  
               "asm_roll": false,  
               "openssl_crypto": true,  
               "asm_MD5": false  
           },  
           "checksum_list": [  
               "md5",  
               "md4",  
               "sha1",  
               "none"  
           ],  
           "compress_list": [  
               "zlibx",  
               "zlib",  
               "none"  
           ],  
           "daemon_auth_list": [  
               "sha512",  
               "sha256",  
               "sha1",  
               "md5",  
               "md4"  
           ],  
           "license": "GPLv3",  
           "caveat": "rsync comes with ABSOLUTELY NO WARRANTY"  
       },  
       "ssh": "OpenSSH_9.7p1, OpenSSL 3.0.13 30 Jan 2024",  
       "sshfs": "3.7.3",  
       "encfs": "(no encfs)",  
       "shell": "/bin/bash",  
       "shell-version": "GNU bash, version 5.2.26(1)-release (x86_64-pc-linux-gnu)"  
   }  
}
```

**Comment by buhtz** [comment]
Thank you for this detailed report.

I wonder if it might be more secure to separate the "additional rsync options" between take snapshot & remove. Maybe an extra dialog window offering three fields: "additional rsync options..." 1) for all rsync operations 2) for take snapshot only 3) for remove snapshots.

I wouldn't recommend to run rsync your self as you suggested.

Anyway. It wonders me that setting the environment variable has no effect. To my knowledge this should work. RSYNC_MAX_ALLOC is supported by rsync since version 3.2.2. You are at 3.3.0.

What about your error message when using the environment variable? Do you still see "exceeded --max-alloc=1073741824 setting" or something like "exceeded --max-alloc=5G setting"?

Can you run BIT with "--debug" please. There you can find the real rsync-command used for removing snapshots. I couldn't find an indication in the source code but it might be that somewhere else --max-alloc is explicit set and that is why the environment variable is ignored.

**Comment by xelvinc** [comment]
Thanks for your quick reply. I'll run BiT with "--debug" later today. I've got the snapshot removed in the end by repeatedly removing it until it was completely removed, so I'm not sure if I can reproduce myself anymore.

**Comment by xelvinc** [comment]
Output with --debug:
```
DEBUG: [qt/app.py:1545 Execute.run] Call command "rsync -a --delete -s /tmp/tmpaw6c8_5y/ /mnt/databackup/backintime/backintime/gentoo/root/1/20240202-143318-346"
WARNING: [qt/app.py:1545 Execute.run] Command "rsync -a --delete -s /tmp/tmpaw6c8_5y/ /mnt/databackup/backintime/backintime/gentoo/root/1/20240202-143318-346" returns 22 | [generator] exceeded --max-alloc=1073741824 setting (file=ifuncs.h, line=30)
rsync error: error allocating core memory buffers (code 22) at util2.c(80) [generator=3.3.0]
ERROR: [common/snapshots.py:642 remove] Last rsync command failed with return code "22". See previous WARNING message in the logs for details.
```
I double check the environmental variable with the `env` command, it looks correct:
`RSYNC_MAX_ALLOC=5g`
Also tried with:
`RSYNC_MAX_ALLOC=5368709120`
and with:
`RSYNC_MAX_ALLOC=0`
All with the same result.

Could it be that the python script is running in it's own environment and therefor the systems environment variables are ignored?

**Comment by buhtz** [comment]
> Could it be that the python script is running in it's own environment and therefor the systems environment variables are ignored?

Yes, that is the case. I always forget that. For your workaround open the starter scripts `/usr/bin/backintime` and `/usr/bin/backintime-qt`. One way would be to remove the `-E` option. But I am not sure about side effects.

I tried to reproduce the behavior with that script:

```
#!/usr/bin/env python3
# foo.py
import os
print(os.environ)
```

And calling it:

    $ python3 -Es foo.py

But I still see all (?) environment variables. I would expect I wouldn't see them. I am confused.

**Comment by xelvinc** [comment]
Same here. I'm also not sure if this is the way to go. It might add additional challenges to get it working with cron, but I'm not sure.
[]https://cronitor.io/guides/cron-environment-variables

Your proposal with the 3 fields sounds good to me.

**Comment by buhtz** [comment]
Another short term workaround for you would be to directly edit tools.py in your installation and add the extra option.

https://github.com/bit-team/backintime/blob/8c7afef81ae79be491a848400261eabdc3e90c8c/common/tools.py#L1107

**Comment by xelvinc** [comment]
Yes, that worked. Thanks for the hint.

**Comment by buhtz** [comment]
Because of #2060 I would like to modify my suggested approach:

> I wonder if it might be more secure to separate the "additional rsync options" between take snapshot & remove. Maybe an extra dialog window offering three fields: "additional rsync options..." 1) for all rsync operations 2) for take snapshot only 3) for remove snapshots.

It might be also relevant when it comes to restoring snapshots.

So the extra dialog should offer "additional rsync options..." for
1) all rsync operations
2) take snapshot only
3) remove snapshots
4) restore snapshots

**Comment by xelvinc** [comment]
Makes sense. If I understand correctly, the final options are than always 1 + one of the 3 others, depending on the action?

**Comment by buhtz** [comment]
I would say, yes. But I haven't fully thought this through yet.

*EDIT*: See #2141 as follow-up.

**Comment by dongbohu** [comment]
Does it make more sense to simply call `rm -rf ...` command to remove an old snapshot?

**Comment by buhtz** [comment]
> Does it make more sense to simply call `rm -rf ...` command to remove an old snapshot?

I am not the original founder of that code. I was not able to find documentation or comments about the reason why rsyn was choosen in favor of rm. But my hypothesis is that rsync is much faster. The second point might be the more detailed error handling and error feedback, compared to rm.

**Comment by dongbohu** [comment]
@buhtz Thank you for your reply. 

I don't know how much faster `rsync` is than `rm`; but when using `rm` command, the config will be simpler (neither the maintainers nor the users need to worry about extra `rsync` options when removing old snapshots).

**Comment by buhtz** [comment]
> I don't know how much faster `rsync` is than `rm`;

Me either. It is a theoretical conclusion based on my knowledge how rsync and cp works. Rsync will be much faster with more files, directories and deep directory trees, which is the case in backups. The "smart remove" feature (aka Remove & Retention) will also remove multiple backups.

I don't have resources to make real life tests. But BIT is 17 years old and have seen many other developers. I do trust their experience.

> but when using `rm` command, the config will be simpler (neither the maintainers nor the users need to worry about extra `rsync` options when removing old snapshots).

That is really a very good argument. I love simplicity and usually try to modify BIT in a way to make it (or its code base) simpler to lower maintenance burden. But in this special case I wouldn't touch the existing feature.

**Comment by Rudranarayan666** [comment]
Hi, I'd like to work on this issue.

Before I start implementing it, I'd like to confirm the proposed direction.

My understanding is to keep the fix small and avoid any GUI/config changes for now:

* Add a helper in `tools.py` to identify the user-provided rsync options that are safe to apply across rsync operations, such as `--max-alloc`, `--timeout`, `--contimeout`, and `--bwlimit`.
* Pass these safe options to `tools.rsyncRemove()`, so options configured by the user are also respected when removing snapshots.
* Keep sync-specific options such as `--checksum`, filters, `--link-dest`, `--delete*`, etc. out of the removal command to avoid changing or breaking its existing behaviour.
* Add tests to verify that safe options are propagated while sync-only options are not.

This should address the reported `--max-alloc` issue while keeping the scope focused. It also shouldn't conflict with the broader three-field UI/config proposal discussed in #1984 / #2141, which could be implemented separately later.

Does this approach look reasonable, or is there anything you'd like me to adjust before I start working on it?

Thanks!

**Comment by buhtz** [comment]
The issue seems to complex for a first-time contributor. I would say no. Your post and you profile are smell much of AI. Are you able to understand BIT and your modifications? Are you really able to take full responsibility of code you submit? This project is not a beginner playground.

It might also be that you missed the information intended to be for new contributors, so 
please read this FAQ entries:
 * [Why do I need to introduce myself?](https://github.com/bit-team/backintime/blob/dev/FAQ.md#why-do-i-need-to-introduce-myself)
 * [Can you assign this to me?](https://github.com/bit-team/backintime/blob/dev/FAQ.md#can-you-assign-this-to-me)
 * [Can I use @ mentions freely in issues or PRs?](https://github.com/bit-team/backintime/blob/dev/FAQ.md#can-i-use--mentions-freely-in-issues-or-prs)
 * [Can I contribute without using the software?](https://github.com/bit-team/backintime/blob/dev/FAQ.md#can-i-contribute-without-using-the-software)
 * [Can I boost my commit count?](https://github.com/bit-team/backintime/blob/dev/FAQ.md#can-i-boost-my-commit-count)
 * [Can I submit AI-generated contributions?](https://github.com/bit-team/backintime/blob/dev/FAQ.md#can-i-submit-ai-generated-contributions)


**EDIT**: I don't support the approach your "AI" hallucinated. It is clearly stated in the issues that #2141 need come first, and before that the issues blocking #2141.

**EDIT2**: If you really want to join this project I am willing to guide and mentoring you. But I need feedback to my questions.
